src/command.rs

```diff
@@ -428,30 +428,226 @@ pub trait Parameters: Sized {
+                    .and_then(|x| x.as_string())
+                    .map(|x| x.to_owned());
+                if let Some(version_cond) = browser_version {
+                    if version_cond.as_str() != "*" {
+                        let found_version = browser_capabilities.browser_version(&merged);
+                        if let Some(version) = found_version {
+                            if !browser_capabilities.compare_browser_version(&*version,
+                                                                             &*version_cond)
+                                .unwrap_or(false) {
+                                return None;
+                            }
+                        } else {
+                            return None;
+                        }
+                    }
+                }
+                let platform_name = merged.get("platformName")
+                    .and_then(|x| x.as_string())
+                    .map(|x| x.to_owned());
+                if let Some(name) = platform_name {
+                    if name.as_str() != "*" &&
+                        Some(name) != browser_capabilities.platform_name(&merged) {
+                            return None;
+                        }
+                }
+                let platform_version = merged.get("platformVersion")
+                    .and_then(|x| x.as_string())
+                    .map(|x| x.to_owned());
+                if let Some(version) = platform_version {
+                    if version.as_str() != "*" &&
+                        Some(version) != browser_capabilities.platform_version(&merged) {
+                            return None;
+                        }
+                }
+                let accept_insecure_certs = merged.get("acceptInsecureCerts")
+                    .and_then(|x| x.as_boolean())
+                    .unwrap_or(false);
+                if accept_insecure_certs &&
+                    !browser_capabilities.accept_insecure_certs(&merged) {
+                    return None;
+                }
+
+                return Some(merged)
+            })
+            .next()
+    }
+}
+
+#[derive(PartialEq)]
+pub struct LegacyNewSessionParameters {
+    pub desired: BTreeMap<String, Json>,
+    pub required: BTreeMap<String, Json>,
+}
+
+impl CapabilitiesMatching for LegacyNewSessionParameters {
+    fn match_browser<T: BrowserCapabilities>(&self, browser_capabilities: &mut T)
+                                             -> Option<BTreeMap<String, Json>> {
+        /* For now don't do anything much, just merge the
+        desired and required and return the merged list. */
+
+        let mut capabilities: BTreeMap<String, Json> = BTreeMap::new();
+        self.required.iter()
+            .chain(self.desired.iter()
+                   .filter(|&(key, _)| !self.required.contains_key(key)))
+            .fold(&mut capabilities,
+                  |mut caps, (key, value)| {caps.insert(key.clone(), value.clone()); caps});
```

Comment:
I think `fold` is considered the normal term in functional programming circles. 

I guess I can replace it with the second form if you prefer (fwiw I find it slightly more confusing because it depends on the order of iteration).